Pure other import

ajkiessl opened the pull request:
This includes mappers for importing "other" publications from Activity Insight and Pure.  This also includes the mapper for exporting to Orcid.  The mappers are all implemented and the code that was preventing "other" pub imports has been removed.  This should probably be tested in a staging env before pushing to prod.

EricDurante commented:
@ajkiessl 
That's a good question, but I don't know if I can answer it without a more in-depth analysis of the data. I also don't know if I should be the person to make that call. The decision to only import "articles" from various sources in the first place wasn't really mine (though it was probably somewhat informed by me). I suppose that a lot of these decisions get somewhat deferred to me because of my (relative) familiarity with the data and the overall system, but ultimately I'm not a PSU librarian.

Here's what I _can_ tell you:

Just taking a glance at a sampling of the publications types from the WoS data, I see things like:

- Article
- Book
- Biographical-Item
- Biography
- Proceedings Paper
- Meeting
- Book Review
- Letter
- Abstract
- Meeting Abstract
- Book Chapter
- Editorial
- Review
- News Item
- News
- Editorial Material
- Correction

I don't know of a good way to produce an exhaustive list of all of the possible types. I also don't know if the structure of the metadata is the same for all of these other publication types and if we could identify the publications and their associations (like authors) in the same way that we do for articles.

Another thing to consider is that because that dataset is so vast and because we just have access to it via a physical disk, it's really somewhat cumbersome to get it imported (or reimported) into RMD. Actually, it's almost physically impossible (at least for me) to do that at this point since I'm limited by my internet connection.

The takeaway, I guess, is that there are major unknowns and obstacles with regard to importing more WoS data, and I really don't know how much value it would add and whether or not it would be worth it.

ajkiessl commented:
@EricDurante  Ok I'll put that aside for now.

I'm going to test these integrations out locally.  I should have a Pure API key somewhere that works.  I'll let you know how it goes.

EricDurante reviewed (commented):
I haven't dug way into the details here, but I think this generally looks good.

Before this can go into production, it will need to be accompanied by some changes to the business logic in the rest of the app. We'll need to define the idea of "article" and "non-article" publications for the purposes of RMD workflows and then make sure that each behaves correctly in terms of how they're listed/presented in various places and which workflows (open access, ORCiD, etc.) are applicable to them.

ajkiessl commented:
@EricDurante  A few things I've observed regarding this ☝️ :

- Most of the "other" publications pulled from Pure have DOIs.  The opposite is true for the ones pulled from AI.  Regardless, both of these should still work with the OA button import.
- The ORCiD export is being handled by the export mapper that I added in this PR.  So exporting pubs to ORCiD should not be affected.
- Publication type is displayed in the deduping process.  I don't expect there to be any major effects to the way admins merge records.  They will have to judge publication type in their decision to merge more harshly now.
- API should be unaffected

The only thing I can think of that might need to be added is to display the publication type in the user profile and the sample public profile.  Though, there may be things that I have not considered.

ajkiessl commented:
I guess one major issue that may come up is the Open Access notification going out to faculty who have a publication that is not a journal article.  They might not have that.  So we might want to stick to journal articles and derivations of journal articles for the OA emails.

EricDurante commented:
@ajkiessl Yeah, I think we'll have to prevent the whole open access workflow from applying to "non-article" publications, so we might need to reconsider how the publication list that you find [here](https://metadata.libraries.psu.edu/profile/publications/edit) should work. There's actually quite a lot to consider, I think. I definitely don't have all of the answers either, so we might all need to discuss this together. For a lot of use cases, I think that "article" and "non-article" really need to be two separate lists even though they are all "publications". I don't think that displaying all of the different kinds of publications everywhere and just noting the type of each is necessarily going to be desirable.

One thing to consider would be how we want the [publications-by-user API endpoint](https://metadata.libraries.psu.edu/api_docs/swagger_docs/v1#!/user/findUserPublications) to behave vs. the [profile-by-user API endpoint](https://metadata.libraries.psu.edu/api_docs/swagger_docs/v1#!/user/findUserProfile) with regard to non-article publications. For the publications endpoint, we can definitely just list all publications and include the type. For the profile endpoint, we might want to either exclude "non-article" publications altogether, or we might want to include them as a separate list in addition to the list of "articles". My sense is that we wouldn't want them in the same list. Take the example of the Huck Institute's website:  They definitely only want to list academic journal articles on their faculty profile pages. Now, they're actually using the "publications" API rather than the "profile" API, so they can slice and dice the data however they want, but if they _were_ using the "profile" API, then they'd expect it to only list journal articles.

If it is the case that we need to separate publication types in the public profiles that we provide, then we would likewise not want all publications in the same list in the profile management interface. In any case, we wouldn't want "non-articles" to behave the way that "articles" currently do in that list. For example, we wouldn't want to display any open access status information, or display the title as a link if open access info is missing, or allow a user to complete any part of the open access workflow (i.e. getting an email, adding a URL, visiting Scholarsphere, adding a waiver, etc.) for "non-articles".

Do you agree with that assessment?

ajkiessl commented:
@EricDurante  Ok, so can we confirm this statement: "Whether or not any individual faculty member will want non-article publications in their profile **is arbitrary**"?  If that is the case, then I agree with the assessment.  I was under the impression that the faculty were upset that their non-article publications were **not** showing up in their profiles (perhaps this was just the case for a select few).  I may have misinterpreted that.  I also agree that this might be a better question for the broader group.

EricDurante commented:
@ajkiessl Yeah, that's actually a slightly different question, I think, and I don't really know the answer. What I'm drawing on to form these thoughts are the couple of specific use cases that we know about. 

We know that we have a case with the College of IST where some faculty were expecting to see at least some (but perhaps not all?) types of their non-journal-article publications listed along with their journal articles (you're definitely right about that), and the folks in charge of the website where that would be displayed seemed to be fine with that. In that case, I don't know if the website developers would be using the "publications" API or the "profile" API.

Then we have the case of the Huck Institute where the point of their faculty profiles is really to showcase their peer-reviewed research and how impactful it is. Their site assumes that it's only listing journal articles that are peer reviewed and cited by other articles, etc. If we were providing the exact list of publications for them to show, we know for sure that they wouldn't want books or conference proceedings or blog posts or anything else mixed into that list. However, we know that in this case they're using the "publications" API and not the "profile" API, so they can filter and sort the list however they want.

We also have the college of Ag Sci web developer who I think is also using the "publications" API also and doesn't seem to be thinking about what kinds of publications they're getting (which means that they're fine with what they're getting right now, I guess).

What this comes down to really is trying to understand if/how people are going to use these API features and then deciding what we should provide and how flexible it should be. So there's no issue with the "publications" API - we just provide all of the data and people can use it however they want. There's already a decent amount of flexibility built into the "profile" API too. People can choose on a publication-by-publication basis whether or not they want a publication to show up in their profile, so to avoid a jarring change, at the very least we might want to make non-article publications hidden by default so that people can turn them on only if they want to. However, that doesn't help if what people really want is to see all of their research articles in one list, and say, all of their books (or all of their non-article pubs of any type, etc.) in a separate list in their profile.

So yeah, I think the decision of _whether or not_ to show certain pubs could be arbitrary for any _individual_ person, and so could the decision about _how_ to show/organize them. These decisions could also be arbitrary for a whole website rather than individual people.

There's only so much flexibility that we can reasonably allow, so the answer for the individual might be, "Sorry, if your department website is using the profile API, then you can't list your different types of publications in separate groups because most people don't want that, but you can choose which ones you want to show, and only journal articles will be shown by default". Or if there's a strong indicator that most people would want two separate lists, then maybe we should do that, and we'd have to say, "Sorry.." to the few folks who want just one list (or they could ask their web developers to use the more flexible API instead).

In the absence of a lot of concrete information about what we should do here, I think we'll just need to discuss with the whole team and decide on something as a best guess.

ajkiessl commented:
@EricDurante  Ok, thanks for explaining this Eric.  I agree with all of this, and the previous assessment. There were parts of the OA workflow I was not considering.  Still trying to wrap my head around all of the bits and pieces.

EricDurante commented:
@ajkiessl There are a lot of bits and pieces, and I'm very happy to help you get a grasp on all of them so that I'm not the only one! 🙂  This whole effort will turn out way better if there's more than one brain fully engaged with all of the details.

EricDurante commented:
@ajkiessl Do you feel like you have a good handle on what else needs to be done for this now that we've made some decisions? It sounds like our key takeaways are:

- We'll keep all types of publications in the same list in the faculty profiles, but we may need to change the presentation a bit to accommodate the various new types. This presentation change might be necessary for the faculty-facing profile management interface and for the public-facing profile pages and the API endpoint/example profile site. We'll need to look at this in more detail.
- We'll adjust all aspects of the open access workflow to exclude all types of publications to which the open access policy does not apply.
- We'll double-check and make sure that we're properly listing the publication type in our API endpoints that provide publication metadata, and it would probably be useful for our API users if we documented those types.

There's one other thing that I thought of too:
- The Swagger documentation in the Publication model is written in such a way that it kind of assumes that all publications are journal articles, so we'll need to tweak that wording a little bit in places and regenerate that documentation.

Be on the lookout for anything else that might need to change that I might be forgetting. 

I can try to find some time to do some pairing with you on this if you'd like.

ajkiessl commented:
@EricDurante Yep, this all sounds good.  I'll get to work on this early this week and have something put together before we pair.  I'm free tomorrow afternoon, Wednesday morning, and pretty much all day Thursday and Friday.

EricDurante commented:
@ajkiessl Awesome - I have pretty much the same availability as you this week, so take however much time you need, and then just let me know when you'd like to look at stuff together.

ajkiessl commented:
@EricDurante I _think_ I've covered all of the requirements and tests for this.  

- The Ai and Pure importers now pull all pub types using the mappers
- Faculty profile page has a separate tab for other publications
- The public profile is the same way
- The demo site + example code is updated to mimic this
- The profile API endpoint also separates out other_publications (uses JSON object to organize different pub types)
- Updated swagger docs to represent this
- The publications endpoint was *not* changed
- Any OA workflow (including the email notifications) exclude non-article publications
- Added/changed tests to test suite for all of this

I am currently running the AI and Pure imports one more time to see if there are any issues.  Other than that I think this is ready to go?  

Also, is there any other documentation I should add?  Maybe something to briefly explain that there are two main types of publications (non-article and article) that behave differently.

EricDurante reviewed (changes requested):
@ajkiessl
Other than my two comments that ask for minor fixes, I think this looks good. 

I've only reviewed the code changes here, so I haven't attempted a new data import, nor have I done any manual testing/QA on this, but it sounds like you have those things pretty well covered.

Let me know if you have any questions, or if you'd like to look at anything else together.

Nice work!

EricDurante commented:
@ajkiessl 
Oh, and to your question about adding documentation generally - Yes, it would be great to do that if you can, but again, there's no need to hold up this merge for it. If you'd like to just come back around and do that whenever you have a chance, that would be fine with me.

ajkiessl commented on app/assets/javascripts/profile/manage.js.coffee:
```diff
@@ -33,4 +33,13 @@ $(document).on('turbolinks:load', ->
+        data: $(this).sortable('serialize')
+      )
+  )
 )
```
Since the order of the other-publications' presentation in the public profile is bound to the default order of the publication types then maybe it would be best to just remove the ability to order the other-publications for now.   They seem to be less important to the faculty anyways.  If someone requests the ability to order them, then we can work on it then.  I'll add a github issue to consider it.

ajkiessl commented:
@EricDurante I ran another Pure import and it works.  I'm going to run another with my local data copied from the current production data just to confirm.  I ran the Activity Insight import but it got interrupted at about 95% with a network timeout.  I think it's fair to say that import works, though (assuming it doesn't get interrupted).  Combined, the two integrations add about 100,000 new publications with about 18,000 duplicate groupings.

EricDurante commented on app/models/publication.rb:
```diff
@@ -62,10 +62,14 @@ def self.publication_types
 
-  scope :subject_to_open_access_policy, -> { where('published_on >= ?', Publication::OPEN_ACCESS_POLICY_START) }
+  scope :subject_to_open_access_policy, -> { where("published_on >= ? AND publication_type ~* 'Journal Article'", Publication::OPEN_ACCESS_POLICY_START) }
 
```
I _think_ you can use the new scope inside of this lambda too. For example, 
```ruby
-> { journal_article.where('published_on >= ?', Publication::OPEN_ACCESS_POLICY_START) }
```
... or did that not work?

EricDurante commented:
@ajkiessl See my one comment/question, but these changes look great.

How did the imports in your dev env look? We better warn someone before we drop the 18,000 new duplicate publications bomb 💣...

I'm pretty swamped this week, but I'd like to try to help you with anything else that you need here. Let me know if there's anything else specifically that you'd like me to take a look at. Otherwise, if things are working well for you in development, I'd say let's merge this.

EricDurante commented:
@ajkiessl That sounds fine to me. The Pure import isn't running in production automatically yet, but the Activity Insight import is running every evening at 10:00, I believe. So whenever you merge/deploy, you can expect the non-article publications from Activity Insight to be imported and for the duplicate groups to be created that same night.
